### PR TITLE
drivers: flash: Add xmc4xxx flash drivers

### PR DESCRIPTION
**Pull request by talih0**
Add xmc4xxx flash drivers.

Signed-off-by: Andriy Gelman <andriy.gelman@gmail.com>

### CODE REVIEW

File: drivers/flash/Kconfig.xmc4xxx [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+config SOC_FLASH_XMC4XXX
+	bool "XMC4XXX flash driver"
+	default y
+	depends on DT_HAS_INFINEON_XMC4XXX_FLASH_CONTROLLER_ENABLED
+	select FLASH_HAS_PAGE_LAYOUT
+	select FLASH_HAS_DRIVER_ENABLED
+	help
+		Enables XMC4XXX flash driver.
```
Comment (mbolivar-nordic) [inline drivers/flash/Kconfig.xmc4xxx:11]: Use `<TAB><SPACE><SPACE>` instead.

File: dts/bindings/flash_controller/flash-pages-layout.yaml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+include: base.yaml
+
+compatible: "flash-pages-layout"
```
Comment (mbolivar-nordic) [inline dts/bindings/flash_controller/flash-pages-layout.yaml:3]: A generic binding like this being added for a specific SoC without any justification in the commit message? No way. Either justify this or do something specific in your own binding, like a `child-binding` or additional array properties.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]

File: dts/bindings/flash_controller/flash-pages-layout.yaml (hunk 2 above)
Comment (talih0) [reply dts/bindings/flash_controller/flash-pages-layout.yaml]: thanks, I updated to use child-binding

File: dts/arm/infineon/xmc4xxx.dtsi [hunk 3]
```diff
@@ -33,8 +33,15 @@
...
 	};
 
-	flash0: serial-flash@c000000 {
-		compatible = "serial-flash";
+	flash_controller: flash_controller@c000000 {
+		compatible = "infineon,xmc4xxx-flash-controller";
+		erase-value = <0>;
+		#address-cells = <1>;
+		#size-cells = <1>;
+		flash0: flash@c000000 {
```
Comment (mbolivar-nordic) [inline dts/arm/infineon/xmc4xxx.dtsi:41]: This looks incorrect to me. The flash controller and the flash memory itself have the exact same base address? That can't be right, can it? Line 41 here is saying that the first addressable byte of flash storage is at address 0xc000000. But line 36 above is saying that the base address of the register block where you can control the flash itself (like write registers to erase pages, etc.) is the same.

File: drivers/flash/soc_flash_xmc4xxx.c [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(offset);
+	ARG_UNUSED(size);
+
+	return -ENOTSUP;
+}
+#endif
+
+const struct flash_parameters *flash_xmc4xxx_get_parameters(const struct device *dev)
```
Comment (mbolivar-nordic) [inline drivers/flash/soc_flash_xmc4xxx.c:174]: Should be static, no?

File: drivers/flash/soc_flash_xmc4xxx.c [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+}
+
+#define SET_PAGES(node_id)    \
+	{.pages_count = DT_PROP(node_id, pages_count), .pages_size = DT_PROP(node_id, pages_size)},
+
+#if CONFIG_FLASH_PAGE_LAYOUT
+static const struct flash_pages_layout flash_xmc4xxx_pages_layout[] = {
+	DT_FOREACH_CHILD(DT_NODELABEL(pages_layout), SET_PAGES)};
+
+void flash_xmc4xxx_page_layout(const struct device *dev, const struct flash_pages_layout **layout,
```
Comment (mbolivar-nordic) [inline drivers/flash/soc_flash_xmc4xxx.c:44]: static?

File: dts/bindings/flash_controller/infineon,xmc4xxx-flash-controller.yaml [hunk 6]
```diff
@@ -0,0 +1,25 @@
+description: XMC4XXX flash controller
+
+compatible: "infineon,xmc4xxx-flash-controller"
+
+include: flash-controller.yaml
```
Comment (mbolivar-nordic) [inline dts/bindings/flash_controller/infineon,xmc4xxx-flash-controller.yaml:5]: Like the DTS, the binding seems to be mixing up the flash memory and the flash controller.

- `erase-value` looks like it should be a property of the flash memory, not the controller, but I see we are making the same mistake in zephyr,sim-flash.yaml, so it's OK to leave this as-is
- however, the flash page layout is definitely an attribute of the flash memory itself, not the controller, right? I think this should be handled in the flash then.

What I recommend is doing this:

- creating a new file, `dts/bindings/mtd/infineon,xmc4xxx-nv-flash.yaml`, which includes `soc-nv-flash.yaml` but uses its own compatible value
- moving the `child-binding` there as a new custom property for your SoC flash binding

Pinging @nvlsianpu as the flash maintainer to confirm this idea is OK.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
Thanks for the update. I see what you're trying to do with the page layout definition now. Here's the next round of review; after this the PR should be OK from my side, unless other major changes come in.

File: dts/bindings/flash_controller/infineon,xmc4xxx-flash-controller.yaml (hunk 6 above)
Comment (nvlsianpu) [reply dts/bindings/flash_controller/infineon,xmc4xxx-flash-controller.yaml:5]: > erase-value looks like it should be a property of the flash memory, not the controller, but I see we are making the same mistake in zephyr,sim-flash.yaml, so it's OK to leave this as-is

Looks like this is only used by the zephyr,sim-flash.yaml. Reason is that we want to have it configurable. For real drivers we embedded this value in its flash driver implementations. And yes - it should be a flash property, although I never seen flashcontroller which control memories with different eras-value.
Is it really configurable for this driver. If not I would remove it.

Comment (nvlsianpu) [reply dts/bindings/flash_controller/infineon,xmc4xxx-flash-controller.yaml:5]: What @mbolivar-nordic suggested sounds ok to me.

**Review by nvlsianpu — CHANGES_REQUESTED** [review]
I would divide only commit there into two separate.
One for DTS patch and the second for the driver implementation (with Kconfig).

Regard erase-value - if needed I recommend to add it to the `dts/bindings/mtd/soc-nv-flash.yaml` in its own commit

File: dts/arm/infineon/xmc4xxx.dtsi (hunk 3 above)
Comment (talih0) [reply dts/arm/infineon/xmc4xxx.dtsi]: yes, good point. I updated the flash controller base address.

File: drivers/flash/soc_flash_xmc4xxx.c (hunk 4 above)
Comment (talih0) [reply drivers/flash/soc_flash_xmc4xxx.c]: yep, thanks.

**Comment by talih0** [comment]
> I would divide only commit there into two separate. One for DTS patch and the second for the driver implementation (with Kconfig).
> 
> Regard erase-value - if needed I recommend to add it to the `dts/bindings/mtd/soc-nv-flash.yaml` in its own commit

sounds good, I've split up the commits. thanks.

**Review by mbolivar-nordic — APPROVED** [review]
DT review comments are addressed. Leaving the flash driver +1 up to subsystem maintainer @nvlsianpu

**Comment by talih0** [comment]
ping

**Review by nvlsianpu — APPROVED** [review]

**Pull request merged**